README.md
# NodeJs Kinesis Producer

<p align="center">
  <img width="150" height="150" src="https://i.imgur.com/DzHBVsI.png" alt="Logo">
</p>

<p align="center">

<img alt="npm (scoped)" src="https://img.shields.io/npm/v/@twaice/node-kinesis-producer">
<img alt="Github CI" src="https://github.com/TWAICE/node-kinesis-producer/workflows/Build/badge.svg">
<a href="https://codecov.io/gh/TWAICE/node-kinesis-producer">
   <img src="https://codecov.io/gh/TWAICE/node-kinesis-producer/branch/main/graph/badge.svg?token=YDL7IV4MTC"/>
</a>    
<img alt="npm" src="https://img.shields.io/npm/dw/@twaice/node-kinesis-producer">
<img alt="Maintained" src="https://img.shields.io/badge/Maintained%3F-yes-green.svg">
<img alt="License" src="https://img.shields.io/github/license/TWAICE/node-kinesis-producer">

</p>

### Description

Kinesis Producer library for NodeJS inspired by [Kiner](https://github.com/bufferapp/kiner).

### Installation

```
$ npm install @twaice/node-kinesis-producer
```

### Usage

```typescript
import KinesisProducer from '@twaice/node-kinesis-producer';

const producer = new KinesisProducer(
    streamName: "stream-name",
    clientConfig: {
        region: "eu-west-1"
    },
    maxRetries: 10
);

const records = [{
    Data: 'test'
}];
await producer.putRecords(records);
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.4](https://github.com/TWAICE/node-kinesis-producer/compare/v0.0.3...v0.0.4) (2021-02-17)

### [0.0.3](https://github.com/TWAICE/node-kinesis-producer/compare/v0.0.2...v0.0.3) (2021-02-16)

### 0.0.2 (2021-02-16)

### Bug Fixes

- export producer interfaces ([b11af54](https://github.com/TWAICE/node-kinesis-producer/commit/b11af54780d97f1caa11c2603c3fa5b5723a66dc))
- remove npmrc ([7a68866](https://github.com/TWAICE/node-kinesis-producer/commit/7a688662339f3bcd7b983eacd72c062d6f687412))

### [0.0.1](https://github.com/TWAICE/node-kinesis-producer/compare/v1.1.0...v0.0.1) (2021-02-16)

### Bug Fixes

- package.json ([e528da1](https://github.com/TWAICE/node-kinesis-producer/commit/e528da182b1be66395d08189a67b83e7cc7c8455))


src/lib/kinesis.producer.spec.ts
import * as crypto from 'crypto';

import { CreateStreamCommand } from '@aws-sdk/client-kinesis';
import anyTest, { TestInterface } from 'ava';
import kinesalite from 'kinesalite';
import sinon from 'sinon';
import { v4 as uuidv4 } from 'uuid';

import { delay } from '../helpers/utils';

import { KinesisRecord } from './interfaces';
import { KinesisProducer } from './kinesis.producer';

const KINESIS_TEST_STREAM = 'test-stream';
const MAX_RETRIES = 3;
const KINESIS_CONFIG = {
  endpoint: 'http://localhost:4567',
  tls: false,
  region: 'eu-west-1',
  credentials: {
    accessKeyId: 'test',
    secretAccessKey: 'test',
  },
};
const test = anyTest as TestInterface<{
  producer: KinesisProducer;
  kinesis: {
    listen: (port: number, callback: unknown) => void;
    close: (callback: unknown) => void;
  };
}>;

test.beforeEach.cb('it initializes test kinesis stream', (t) => {
  t.context.kinesis = kinesalite({ createStreamMs: 50 });
  t.context.kinesis.listen(4567, async (err: Error) => {
    if (err) throw err;

    t.context.producer = new KinesisProducer({
      streamName: KINESIS_TEST_STREAM,
      clientConfig: KINESIS_CONFIG,
      maxRetries: MAX_RETRIES,
      loggingEnabled: false,
    });

    await t.context.producer.client.send(
      new CreateStreamCommand({
        ShardCount: 1,
        StreamName: KINESIS_TEST_STREAM,
      })
    );
    await delay(2000);
    t.end();
  });
});

test.serial('it handles large amount of records', async (t) => {
  for (let i = 0; i < 100; i++) {
    const records = [];
    for (let j = 0; j < 500; j++) {
      records.push({
        Data: JSON.stringify({
          id: uuidv4(),
          attr: Math.random(),
          body: crypto.randomBytes(Math.ceil(Math.random())).toString('hex'),
        }),
      });
    }
    await t.context.producer.putRecords(records);
  }
  t.assert(true);
});

test.serial('it handles multiple large payloads', async (t) => {
  for (let i = 0; i < 10; i++) {
    const records: KinesisRecord[] = [];
    for (let j = 0; j < 10; j++) {
      const singleRecord = [];
      for (let k = 0; k < 1000; k++) {
        singleRecord.push({
          id: uuidv4(),
          attr: Math.random(),
          body: crypto.randomBytes(Math.ceil(Math.random())).toString('hex'),
        });
      }

      records.push({
        Data: JSON.stringify(singleRecord),
        PartitionKey: crypto.randomBytes(20).toString('hex'),
        ExplicitHashKey: `${j}`,
      });
    }
    await t.context.producer.putRecords(records);
  }
  t.assert(true);
});

test.serial('it handles large payload', async (t) => {
  const singleRecord: Record<string, unknown>[] = [];
  for (let k = 0; k < 100000; k++) {
    singleRecord.push({
      id: uuidv4(),
      attr: Math.random(),
      body: crypto.randomBytes(Math.ceil(Math.random())).toString('hex'),
    });
  }
  await t.throwsAsync(async () => {
    await t.context.producer.putRecord({
      Data: JSON.stringify(singleRecord),
    });
  });
});

test.serial('it throws error on empty payload', async (t) => {
  await t.throwsAsync(async () => {
    await t.context.producer.flushQueue();
  });
});

test.serial('it retries failed records', async (t) => {
  const sendFake = sinon.fake();
  const error = await t.throwsAsync(async () => {
    t.context.producer.client.send = sendFake;
    await t.context.producer.putRecords([
      {
        Data: 'test',
      },
    ]);
    await t.context.producer.flushQueue();
  });
  t.is(sendFake.callCount, 3);
  t.is(error.message, `Max retries ${MAX_RETRIES} reached for this batch.`);
});

test.serial('it flushes the queue periodically', async (t) => {
  const flushQueueFake = sinon.fake();
  t.context.producer.flushQueue = flushQueueFake;
  t.context.producer.loggingEnabled = true;
  await t.context.producer.putRecords([
    {
      Data: 'test',
    },
  ]);
  await delay(5000);
  t.assert(flushQueueFake.called);
});

test.serial("it doesn't flush queue if empty", async (t) => {
  const flushQueueFake = sinon.fake();
  t.context.producer.flushQueue = flushQueueFake;
  await delay(5000);
  t.assert(flushQueueFake.notCalled);
});

test.afterEach.cb('it closes test kinesis stream', (t) => {
  t.context.kinesis.close((err: Error) => {
    t.end(err);
    t.context.producer.close();
  });
});


src/lib/interfaces.ts
/**
 * @packageDocumentation
 * @module KinesisProducer
 */

import { KinesisClientConfig } from '@aws-sdk/client-kinesis/KinesisClient';

export interface KinesisRecord {
  Data: Buffer | string;
  PartitionKey?: string;
  ExplicitHashKey?: string;
}

/**
 * streamName: Name of the stream to send the records.
 * clientConfig: Configuration for AWS-SDK Kinesis Client.
 * batchSize: Numbers of records to batch before flushing the queue.
 * batchSizeInBytes: Maximum size in bytes to batch before flushing the queue.
 * batchTime: Maximum of seconds to wait before flushing the queue.
 * maxRetries: Maximum number of times to retry the put operation.
 */
export interface KinesisProducerConfig {
  streamName: string;
  clientConfig?: KinesisClientConfig;
  batchSize?: number;
  batchSizeInBytes?: number;
  batchTime?: number;
  maxRetries?: number;
  loggingEnabled?: boolean;
  logger?: Logger;
}

export interface Logger {
  log(message: unknown, context?: string): unknown;
  error(message: unknown, trace?: string, context?: string): unknown;
  warn(message: unknown, context?: string): unknown;
  debug?(message: unknown, context?: string): unknown;
  verbose?(message: unknown, context?: string): unknown;
}


src/lib/kinesis.producer.ts
/**
 * @packageDocumentation
 * @module KinesisProducer
 */
import * as crypto from 'crypto';

import {
  KinesisClient,
  PutRecordsCommand,
  PutRecordsRequestEntry,
  PutRecordsResultEntry,
} from '@aws-sdk/client-kinesis';

import { delay, getRecordSizeInBytes } from '../helpers/utils';

import { KinesisProducerConfig, KinesisRecord, Logger } from './interfaces';

export class KinesisProducer {
  private readonly streamName: string;
  private queue: Array<PutRecordsRequestEntry>;
  private readonly batchSize: number;
  private readonly batchSizeInBytes: number;
  private readonly batchTime: number;
  private readonly maxRetries: number;
  private queueSizeInBytes: number;
  private _client: KinesisClient;
  private lastFlush: Date;
  private readonly flushIntervalFn: NodeJS.Timer;
  private readonly logger: Logger;
  private _loggingEnabled: boolean;

  /**
   * Kinesis Producer Constructor.
   * @param kinesisProducerConfig: Configuration for the producer.
   */
  constructor(kinesisProducerConfig: KinesisProducerConfig) {
    const {
      streamName,
      clientConfig,
      batchSize,
      batchSizeInBytes,
      batchTime,
      maxRetries,
      logger,
      loggingEnabled,
    } = kinesisProducerConfig;
    this.streamName = streamName;
    this._client = new KinesisClient(clientConfig || {});
    this.queue = [];
    this.batchSize = Math.min(500, batchSize || 500);
    this.batchSizeInBytes = batchSizeInBytes || 1024 * 1024;
    this.batchTime = batchTime || 5;
    this.maxRetries = maxRetries || 3;
    this.queueSizeInBytes = 0;
    this.lastFlush = new Date();
    this.flushIntervalFn = this.flushPeriodically();
    this._loggingEnabled = loggingEnabled != undefined ? loggingEnabled : true;
    this.logger = logger || console;
  }

  /**
   * Add a list of data records to the record queue in the proper format.
   * Convenience method that calls putRecord for each element.
   * @param records: List of KinesisRecord to send.
   */
  public async putRecords(records: Array<KinesisRecord>) {
    for (const record of records) {
      await this.putRecord(record);
    }
  }

  public flushPeriodically(): NodeJS.Timer {
    return setInterval(async () => {
      const currentTime = new Date();
      const diff = Math.abs(currentTime.getTime() - this.lastFlush.getTime());
      if (diff >= this.batchTime) {
        if (this.queue.length > 0) {
          if (this._loggingEnabled) {
            this.logger.warn(
              `Queue not empty and not flushed since ${diff} seconds. Flushing now.`
            );
          }
          await this.flushQueue();
        }
      }
    }, this.batchTime * 1000);
  }

  /**
   * Add a record to batch queue to be sent to kinesis stream.
   * Flushes the queue if flushing conditions are met.
   * @param kinesisRecord
   */
  public async putRecord(kinesisRecord: KinesisRecord) {
    if (!kinesisRecord.PartitionKey) {
      kinesisRecord.PartitionKey = crypto.randomBytes(20).toString('hex');
    }

    if (typeof kinesisRecord.Data == 'string') {
      kinesisRecord.Data = Buffer.from(kinesisRecord.Data);
    }
    const recordSize = getRecordSizeInBytes(kinesisRecord);
    if (recordSize > this.batchSizeInBytes) {
      throw new Error(
        `Record size ${recordSize} cannot be greater than batchSize ${this.batchSizeInBytes}
         for record with PartitionKey: ${kinesisRecord.PartitionKey}`
      );
    }
    const shouldFlush =
      this.queue.length + 1 > this.batchSize ||
      this.queueSizeInBytes + recordSize > this.batchSizeInBytes;

    if (shouldFlush) {
      if (this._loggingEnabled) {
        this.logger.warn(
          `Flushed Queue of Length: ${this.queue.length}, Size: ${(
            this.queueSizeInBytes / 1024
          ).toFixed(2)} KiB`
        );
      }
      await this.flushQueue();
    }
    const putRecordInput: PutRecordsRequestEntry = {
      Data: kinesisRecord.Data,
      PartitionKey: kinesisRecord.PartitionKey,
      ExplicitHashKey: kinesisRecord.ExplicitHashKey,
    };

    this.queue.push(putRecordInput);
    this.queueSizeInBytes += recordSize;
  }

  /**
   * Flush all the current queue records to Kinesis.
   */
  public async flushQueue() {
    await this.sendRecords(this.queue);
    this.lastFlush = new Date();
  }

  /**
   * Send records to the Kinesis stream.
   * Failed records are sent again with an exponential backoff.
   * @param records: Records to put in the stream.
   * @param attempt: Number of times the records have been sent without success.
   * @private
   */
  private async sendRecords(
    records: Array<PutRecordsRequestEntry>,
    attempt = 1
  ): Promise<PutRecordsResultEntry[] | undefined | null> {
    if (attempt > this.maxRetries) {
      throw new Error(`Max retries ${this.maxRetries} reached for this batch.`);
    }

    // exponential backoff before retrying.
    await delay(Math.pow(2, attempt) * 0.1);

    const command = new PutRecordsCommand({
      StreamName: this.streamName,
      Records: records,
    });

    this.resetQueue();

    let response, failedRecordCount;
    try {
      response = await this._client.send(command);
      failedRecordCount = response['FailedRecordCount']
        ? response['FailedRecordCount']
        : 0;
    } catch (error) {
      if (this._loggingEnabled) {
        this.logger.error(
          `Failed to putRecords into kinesis stream: ${error.name}: ${error.message}`
        );
      }
      return await this.sendRecords(records, attempt + 1);
    }

    const failedRecords: PutRecordsRequestEntry[] = [];
    if (failedRecordCount > 0) {
      if (this._loggingEnabled) {
        this.logger.warn('Retrying Failed Records');
      }
      response['Records']?.forEach((record, idx) => {
        if (Object.prototype.hasOwnProperty.call(record, 'ErrorCode')) {
          failedRecords.push(records[idx]);
        }
      });

      return await this.sendRecords(failedRecords, attempt + 1);
    }

    return response['Records'];
  }

  /**
   * Reset the queue.
   * @private
   */
  private resetQueue() {
    this.queue = [];
    this.queueSizeInBytes = 0;
  }

  public close() {
    clearInterval(this.flushIntervalFn);
  }

  get client(): KinesisClient {
    return this._client;
  }

  set client(KinesisClient) {
    this._client = KinesisClient;
  }

  set loggingEnabled(value: boolean) {
    this._loggingEnabled = value;
  }
}
